core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -353,10 +353,15 @@ class DAGScheduler(
     if (mapOutputTracker.containsShuffle(shuffleDep.shuffleId)) {
       val serLocs = mapOutputTracker.getSerializedMapOutputStatuses(shuffleDep.shuffleId)
       val locs = MapOutputTracker.deserializeMapStatuses(serLocs)
-      for (i <- 0 until locs.length) {
-        stage.outputLocs(i) = Option(locs(i)).toList // locs(i) will be null if missing
+      var numAvailableOutputs = 0
+      var i = 0
+      while (i < locs.length) {
+        if (locs(i) ne null) {
+          // locs(i) will be null if missing
+          stage.addOutputLoc(i, locs(i))
```

Comment:
+1 for readability.  A balance is to do foreach on a range `(0 until locs.length).foreach`, which is almost as readable as foreach on the collection and almost as fast as the while loop.  https://github.com/scala/scala/commit/4cfc633fc6